src/main/java/seedu/address/storage/Authentication.java

```diff
@@ -0,0 +1,202 @@
+     */
+    public void lock() throws ZipException {
+        ZipFile dataZip;
+        try {
+            //If ClientBook is locked, encrypt the zip file using the password.
+            if (this.password.isPresent()) {
+                dataZip = new ZipFile(this.getZipPath(), this.password.get().toCharArray());
+                ZipParameters parameters = new ZipParameters();
+                parameters.setEncryptFiles(true);
+                parameters.setEncryptionMethod(EncryptionMethod.AES);
+                parameters.setAesKeyStrength(AesKeyStrength.KEY_STRENGTH_256);
+                parameters.setCompressionLevel(CompressionLevel.NORMAL);
+                parameters.setCompressionMethod(CompressionMethod.DEFLATE);
+                dataZip.addFile(this.filePath.toString(), parameters);
+
+            //ClientBook is not locked, zip the data .json file without a password.
+            } else {
+                dataZip = new ZipFile(this.getZipPath());
+                dataZip.addFile(this.filePath.toString());
+            }
+        } catch (ZipException e) {
+            logger.info("Failed to compress data .json file into a zip file");
+            throw e;
+        }
+
+
+
+
+    }
+
+    /**
+     * Sets a hook to lock the data .json file in the zip file when the application closes.
+     */
+    public void setShutDownHook() {
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            try {
+                this.lock();
+                this.deleteJson();
+            } catch (ZipException e) {
+                e.printStackTrace();
+            }
+        }));
+    }
+
+    public String getPassword() {
+        return this.password.orElse("");
+    }
+
+    public boolean isPasswordPresent() {
+        return this.password.isPresent();
+    }
+
+    public void setPassword(String password) {
+        this.password = Optional.of(password);
+    }
+
+    public void removePassword() {
+        this.password = Optional.empty();
+    }
+
+    private void deleteJson() {
+        File dataJson = new File(this.filePath.toString());
+        dataJson.delete();
+    }
+
+    private boolean attemptUnzip(String password) throws ZipException {
+        ZipFile dataZip = new ZipFile(this.getZipPath());
+        //First check if zip file encrypted and set the password to be used.
+        try {
+            if (dataZip.isEncrypted()) {
+                dataZip.setPassword(password.toCharArray());
+            }
+        } catch (ZipException e) {
+            logger.info("Error attempting to check zip file at: " + this.getZipPath());
+            throw e;
+        }
+
+        //Attempt to unzip the file using the provided password.
+        try {
+            dataZip.extractAll(this.getFolderPath());
+            return true;
+        } catch (ZipException e) {
+            logger.info("Error when unzipping file at: " + this.getZipPath());
+            return false;
+        }
+    }
```

Comment:
This method is fine, but are the 2 try-catch blocks are mainly to log different messages? I was thinking whether you can condense it into one try-catch block, and in the catch block, you differentiate the 2 cases using the ZipException that was caught.

Actually I was looking at the docs, and found that they didn't really specify what ZipException each method threw, but on their ZipException class, there is a getType() method, which returns a ZipException.Type. (https://javadoc.io/static/net.lingala.zip4j/zip4j/2.7.0/net/lingala/zip4j/exception/ZipException.Type.html) 

Perhaps you could check if it is possible to differentiate the cases by checking the exception type? If not then your method is completely fine as it is

Reply:
Good point there, I was looking at the same API a few days ago too. The methods ZipFile.isEncrypted() and ZipFile.extractAll() did not specify which type of ZipException was thrown. The try blocks are separated because I was using the second exception to return false, while the first try block was used to catch the error thrown when checking if the zip file is encrypted.

Reply:
I see, since the API doesn't specify then it's difficult to catch the different cases ourselves. In that case then two try-catch blocks is a nice way to circumvent that👍🏼